go/osagent/osagent.go

```diff
@@ -645,3 +745,609 @@ func AbortSeed(seedId string) error {
+// as 'we should have at least' space check, because we can't make any accurate estimations for logical backups
+func GetMySQLDatabaseInfo() (dbinfo MySQLDatabaseInfo, err error) {
+	dbi := make(map[string]*MySQLDatabase)
+	var physicalSize, tokuPhysicalSize, logicalSize int64 = 0, 0, 0
+	databases, err := dbagent.GetMySQLDatabases()
+	for _, db := range databases {
+		engines, err := dbagent.GetMySQLEngines(db)
+		if err != nil {
+			log.Errore(err)
+		}
+		for _, engine := range engines {
+			if engine != "TokuDB" {
+				physicalSize, err = DirectorySize(path.Join(config.Config.MySQLDataDir, db))
+				if err != nil {
+					log.Errore(err)
+				}
+			} else {
+				tokuPhysicalSize, err = dbagent.GetTokuDBSize(db)
+				if err != nil {
+					log.Errore(err)
+				}
+			}
+		}
+		physicalSize += tokuPhysicalSize
+		if config.Config.CompressLogicalBackup {
+			logicalSize = int64(float64(physicalSize) * 0.6)
+		} else {
+			logicalSize = int64(float64(physicalSize) * 0.8)
+		}
+		dbi[db] = &MySQLDatabase{engines, physicalSize, logicalSize}
+		dbinfo.MySQLDatabases = dbi
+	}
+	dbinfo.InnoDBLogSize, err = dbagent.GetInnoDBLogSize()
+	if err != nil {
+		log.Errore(err)
+	}
+	return dbinfo, err
+}
+
+func GetAvailableSeedMethods() []string {
+	var avaliableSeedMethods []string
+	var cmd string
+	for _, seedMethod := range seedMethods {
+		switch seedMethod {
+		case "lvm":
+			cmd = "lvs"
+		case "xtrabackup-stream":
+			cmd = "xtrabackup"
+		default:
+			cmd = seedMethod
+		}
+		err := commandRun(
+			fmt.Sprintf("%s --version", cmd),
+			func(cmd *exec.Cmd) {
+				log.Debug("Checking for seed method", seedMethod)
+			})
+		if err == nil {
+			log.Debug("seed method", seedMethod, "found")
+			avaliableSeedMethods = append(avaliableSeedMethods, seedMethod)
+		}
+	}
+	return avaliableSeedMethods
+}
+
+func StartLocalBackup(seedId string, seedMethod string, databases string) (BackupFolder string, err error) {
+	var cmd string
+	BackupFolder = path.Join(config.Config.MySQLBackupDir, time.Now().Format("20060102-150405"))
+	if err := os.Mkdir(BackupFolder, 0755); err != nil {
+		return "", log.Errore(err)
+	}
+	if !contains(seedMethod, seedMethods) {
+		return "", log.Errorf("Unsupported seed method")
+	}
+	if databases != "" {
+		availiableDatabases, _ := dbagent.GetMySQLDatabases()
+		for _, db := range strings.Split(databases, ",") {
+			if !contains(db, availiableDatabases) {
+				return "", log.Errorf("Cannot backup database %+v. Database doesn't exists", db)
+			}
+		}
+		//if we don't already have mysql database in databases, add it
+		if !strings.Contains(databases, "mysql") {
+			databases += ",mysql"
+		}
+		//and if we are on 5.7 we need to add sys db because sometimes during mysql_upgrade run we can get errors like mysql_upgrade: [ERROR] 1813: Tablespace '`sys`.`sys_config`' exists
+		version, _ := dbagent.GetMySQLVersion()
+		if version == "5.7" && !strings.Contains(databases, "sys") {
+			databases += ",sys"
+		}
```

Comment:
can we append to an array as opposed to concatenating a string?

Reply:
fixed in 3ce673a